New features

BrunoLSousa opened the pull request:
Implementation of new metrics and support for the user to define the directory where he wants to export the .csv files provided by the tool as output.

mauricioaniche commented:
I will review this PR soon, @BrunoLSousa!

BrunoLSousa commented:
Ok, @mauricioaniche! Thank you very much!

mauricioaniche commented on src/main/java/com/github/mauricioaniche/ck/CKVisitor.java:
```diff
@@ -56,21 +56,22 @@ public CKVisitor(String sourceFilePath, CompilationUnit cu, Callable<List<ClassL
 		int modifiers = node.getModifiers();
 		CKClassResult currentClass = new CKClassResult(sourceFilePath, className, type, modifiers);
 		currentClass.setLoc(calculate(node.toString()));
+		
```
Why was this moved here?

I guess the idea of having it before creating the new node is that we notify the current class about a `TypeDeclaration` happening inside of it. Then, we stack a new `CKClassResult`.

BrunoLSousa replied:
I moved this part to get the current class name and set it for the AST, case the code inside the if block is called. If I don't get the current class name before, the program may fail.

mauricioaniche replied:
This is interesting, because it should not make a difference at all. The _CKClassResult_ of the class that was just instantiated is _not_ yet in the stack of classes. It only goes to the stack in now line 77, right? 

So, the `classes.peek().classLevelMetrics.stream().map` that happens inside the `if` you moved, will not go through the newly declared type. 

So, in a way, this code block being here has the same behaviour as the block where it was. What am I missing here?

BrunoLSousa replied:
Yes! You're right, @mauricioaniche. The code block on line 77 now has the same behavior as it did before I moved it. The only reason I changed it was because I need to pass the class name as a parameter to ast now. So this change allows me to take the class name and pass it as a parameter before the ast is instantiated. But in terms of behavior, the code block has the same behavior as before.

mauricioaniche replied:
I guess you will revisit this now that you are going to insert the name of the class during instantiation?

BrunoLSousa replied:
Thank you for your reply, @mauricioaniche! I am revisiting this point, and I will push the changes as soon as possible.

BrunoLSousa replied:
Hi, @mauricioaniche! I have already submitted a new commit with the correction that you had asked. Now, the name of the classes are being inserted during instantiation.

mauricioaniche commented on src/main/java/com/github/mauricioaniche/ck/CKVisitor.java:
```diff
@@ -85,7 +86,7 @@ public boolean visit(TypeDeclaration node) {
 		// there might be class level metrics that use the TypeDeclaration
 		// so, let's run them
-		classes.peek().classLevelMetrics.stream().map(metric -> (CKASTVisitor) metric).forEach(ast -> ast.visit(node));
+		classes.peek().classLevelMetrics.stream().map(metric -> (CKASTVisitor) metric).forEach(ast -> {((ClassLevelMetric)ast).setClassName(className); ast.visit(node);});
```
Setting the class name here, hmm, that also works! What was the reason to do so?

BrunoLSousa replied:
I designed a solution based on the Singleton design pattern for implementing the collection of fan-in and NOC metrics. In the case of fan-in, I have to identify the classes that call a given class. For instance, if I have a class A that calls a class B, I would have to increase 1 for fan-out in A and 1 for fan-in in B. So, I set the class name at this point to identify the class visited at the moment and add it to the fan-in list regarding the classes that it is calling.

In the case of NOC, I do the same thing. But, I observe the call via the keyword **extends** and add the current analyzed class to the NOC list of its superclass.

mauricioaniche replied:
I am now having a better grasp of this `setClassName`. It belongs to the `ClassLevelMetric`. This class represents a metric that is calculated. Setting a class name on it does not really make sense, from a cohesion point of view. The `ClassLevelMetric` is basically an Strategy. Having the name of the class is strange. Also because this means that all strategies will have the name of the current class.

But I understand you need to pass the name of the current class as to then calculate fan in and fan out. 

I'm trying to find a way to avoid this leaking. Do you think it would be better to do it in the `instantiateClassLevelMetricVisitors` method? This method calls the Callable which returns the list of metrics. We'd then manually look for the CBO metric there (look for instanceOf CBO or something) and inject the name of the class only there. It may make this specific method ugly, but it won't hurt the entire metric abstraction...

What do you think?

BrunoLSousa replied:
I think trying to pass the class name to the instantiateClassLevelMetricVisitors method instead of the ClassLevelMetric class interface might be a good solution to avoid this leaking. Do you want me to try implementing this strategy here?

mauricioaniche replied:
I can't recall all the details of my previous review, but let's go for it. It will just hurt a single (factory) method instead of the entire class hierarchy!

BrunoLSousa replied:
Thank you for your reply, @mauricioaniche! I am revisiting this point, and I will push the changes as soon as possible.

BrunoLSousa replied:
Hi, @mauricioaniche! I have already submitted a new commit with the correction that you had asked. Now, the name of the classes are being inserted during instantiation.

mauricioaniche commented on src/main/java/com/github/mauricioaniche/ck/CKVisitor.java:
```diff
@@ -131,7 +134,7 @@ public boolean visit(MethodDeclaration node) {
 		classes.peek().classLevelMetrics.stream().map(metric -> (CKASTVisitor) metric).forEach(ast -> ast.visit(node));
 		if(!classes.peek().methods.isEmpty())
-			classes.peek().methods.peek().methodLevelMetrics.stream().map(metric -> (CKASTVisitor) metric).forEach(ast -> ast.visit(node));
+			classes.peek().methods.peek().methodLevelMetrics.stream().map(metric -> (CKASTVisitor) metric).forEach(ast -> {((MethodLevelMetric)ast).setMethodName(currentQualifiedMethodName); ast.visit(node);});
```
Same question for the method: why setting its name here?

BrunoLSousa replied:
I set the name of the method here because of the same reason I have presented for class.

mauricioaniche replied:
Resolving this one, so we keep the discussion in one comment only.

mauricioaniche reviewed (changes requested):
Hi, @BrunoLSousa! This PR adds lots of important metrics!! Awesome, thanks for that!

I have a few questions just to understand your decisions.

Moreover, this PR currently **misses tests**. You added a lot of metrics. It would be nice to also add tests that show that they work as expected.

See the `tests` folder to see how we do it. Basically, we just create a bunch of simple Java classes in the fixtures folder, call CK there, and then do assertions. Use `CBOTest` as an example!

Would you mind doing that?

BrunoLSousa commented:
> Hi, @BrunoLSousa! This PR adds lots of important metrics!! Awesome, thanks for that!
> 
> I have a few questions just to understand your decisions.
> 
> Moreover, this PR currently **misses tests**. You added a lot of metrics. It would be nice to also add tests that show that they work as expected.
> 
> See the `tests` folder to see how we do it. Basically, we just create a bunch of simple Java classes in the fixtures folder, call CK there, and then do assertions. Use `CBOTest` as an example!
> 
> Would you mind doing that?

BrunoLSousa commented:
> Hi, @BrunoLSousa! This PR adds lots of important metrics!! Awesome, thanks for that!
> 
> I have a few questions just to understand your decisions.
> 
> Moreover, this PR currently **misses tests**. You added a lot of metrics. It would be nice to also add tests that show that they work as expected.
> 
> See the `tests` folder to see how we do it. Basically, we just create a bunch of simple Java classes in the fixtures folder, call CK there, and then do assertions. Use `CBOTest` as an example!
> 
> Would you mind doing that?

Hi, @mauricioaniche! Thank you very much for your feedback! 

As you asked, I have already implemented some test cases regarding the new features and I have added on this PR so that you can analyze.

Regarding your questions, I answered them above.

mauricioaniche reviewed (commented):
Hi, @BrunoLSousa! Sorry for the slow reply here!

I only have one point of discussion. The `setClassName` seems to break the design of the class a bit. I understand that the CBO metric needs it though... See my suggestion in the comments. What do you think? I wanted to avoid adding this `setClassName` to all the metrics and re-setting it everytime we call the loop there...

BrunoLSousa commented:
Hi, @mauricioaniche! Sorry for the slow reply here!

I've seen your comments about my last change on this pull request and I think I am able to implement them here. Could you reopen this pull request so that I can do the changes that you've requested?

Thank you very much!

mauricioaniche commented:
The PR is open again, @BrunoLSousa !

BrunoLSousa commented:
> The PR is open again, @BrunoLSousa !

Thank you very much, @mauricioaniche !

mauricioaniche reviewed (approved):
LGTM!

mauricioaniche commented:
@BrunoLSousa almost there! there are some conflicts! I guess you were working on an old branch? Would you mind fixing them?

BrunoLSousa commented:
> @BrunoLSousa almost there! there are some conflicts! I guess you were working on an old branch? Would you mind fixing them?

@mauricioaniche I have already solved the conflicts.
